Documentation/articles/domain_driven_design/concepts_and_value_objects.md

```diff
@@ -0,0 +1,114 @@
+---
+title: About Concepts and Value Objects
+description: Learn about Concepts and Value Objects in Dolittle
+keywords: domain-driven-design
+author: einari, smithmx, tomasekeli
+---
+
+# Background
+
+Objects which have identity, fall into two distinct categories.  *Entities* and *Value Objects*
+
+## Entities
+
+*Entities* have a unique key, usually a single value but in some cases a composite of 
+multiple values, that identifies the instance of the entity.  When comparing two instances it
+ is sufficient to compare the identities of those entities.
+
+An entity will have the same identifier throughout the lifetime of the system, even though it
+may be represented by different classes and objects hroughout its lifetime.  For example, an
+order in an ecommerce system may at different times be *PlacedOrder*, *PickingList*, 
+*ShippedOrder*, *DelinquentOrder* and so on, but it will always be identified by the same
+*OrderId*.  In other words, it is the value of the key that determines the identity of the entity,
+not the particular class that currently represents it or the current state of that entity.
+
+## Value Objects
+
+*Value Objects* do not have any key that identifies them. A *Value Object* is identified by the
+properties that comprise it.  It follows that two instances of a value object, that have identical
+properties, are considered equal. Examples of value objects could be telephone numbers, a co-ordinate,
+an address, etc..  It should be pointed out that the distinction between entity and value object can
+be domain specific.  What is a value object in one domain, might be an entity in another domain.
+
+
+# Primitive Obsession
+
+The term *primitive obsession* is used to describe the tendency to use language-specific  primitives
+(strings, integers, doubles, etc.) to represent concepts in the domain, rather than fully-fledged types.
+While ultimately all concepts must be expressed in the primitive types available in the language writing
+the domain-code like this leads to a less expressive, less correct domain model. For example, a telephone
+number might be represented as a *string* (or even an integer).  However, there is logic associated with
+a telephone number and aspects of it that are not necessarily captured by a string.  We may be interested
+in area codes, international dialling codes, extensions, country specific formatting rules, and so on.
+Replacing a primitive with a type allows us to capture that while not all strings are telephone numbers, 
+all telephone numbers can be expressed as strings.
+
+{{% notice warning %}}
+Never EVER represent things that look like "numbers" by virtue of being composed solely of digits as integers.
+{{% /notice %}} 
+
+# Value Objects in Dolittle
+
+Value object **MUST** be treated as immutable after creation and initial population.  Dolittle does not 
```

Comment:
We might actually be in a position to enforce this now.  But it's correct as it currently stands...